onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu

```diff
@@ -0,0 +1,775 @@
+                                       const int* d_sorted_indices,
+                                       float* d_logits_in_out,
+                                       float top_p,
+                                       float filter_value,
+                                       int min_tokens_to_keep,
+                                       int batch_size,
+                                       int vocab_size,
+                                       cudaStream_t stream,
+                                       bool is_descending);
+
+template void LaunchFilterLogitsKernel(float* d_sorted_logits_in,
+                                       const int* d_sorted_indices,
+                                       half* d_logits_in_out,
+                                       float top_p,
+                                       float filter_value,
+                                       int min_tokens_to_keep,
+                                       int batch_size,
+                                       int vocab_size,
+                                       cudaStream_t stream,
+                                       bool is_descending);
+
+
+// Ref: https://github.com/pytorch/pytorch/blob/release/1.13/aten/src/ATen/native/cuda/MultinomialKernel.cu
+template <typename scalar_t, typename accscalar_t, unsigned TPB>
+__global__ void sampleMultinomialOnce(int64_t* dest,
+                                      int distributions,
+                                      int categories,
+                                      scalar_t* sampled,
+                                      scalar_t* dist,
+                                      int stride_dist, // dist->stride(0)
+                                      int stride_categories, // dist->stride(1)
+                                      int* d_presence_mask) {
+  using BlockReduce = cub::BlockReduce<float, TPB>;
+  __shared__ typename BlockReduce::TempStorage tmp_storage;
+
+  extern __shared__  unsigned char my_smem[];
+  __shared__ bool found;
+  __shared__ unsigned foundPos;
+  accscalar_t *smem = reinterpret_cast<accscalar_t *>(my_smem);
+  accscalar_t accZero = static_cast<accscalar_t>(0);
+  scalar_t zero = static_cast<scalar_t>(0);
+  for (int curDist = blockIdx.x;
+       curDist < distributions; curDist += gridDim.x) {
+
+    // Each block handles one distribution
+    // First pass, find the total sum of the distribution
+    accscalar_t sum = accZero;
+    scalar_t val;
+    for (int cat = threadIdx.x; cat < categories; cat += blockDim.x) {
+      val = dist[curDist * stride_dist + cat * stride_categories];
+      // CUDA_KERNEL_ASSERT(!at::_isnan(val));
+      // CUDA_KERNEL_ASSERT(!_isinf(val));
+      // CUDA_KERNEL_ASSERT(!(val < zero));
+      sum = sum + static_cast<accscalar_t>(val);
+    }
+    // threadIdx.x == 0 has the sum value from this
+    sum = BlockReduce(tmp_storage).Reduce(sum, cub::Sum()); // sum = cuda_utils::BlockReduceSum(sum, smem);
```

Comment:
sum should be 1.f as the input is SoftMax-ed

Reply:
good catch! we can safely remove this step since this function is only used for topp